Fix Issue 23817 - Reject `is` alias template instance specialization with TemplateParameterList

ntrel opened the pull request:
…mplateParameterList

ntrel committed:
Fix Issue 23817 - `is` alias template instance specialization with TemplateParameterList

ntrel committed:
Fix segfault

thewilsonator commented:
```
../dmd/generated/linux/release/64/dmd -run tools/dspec_tester.d --compiler=../dmd/generated/linux/release/64/dmd
---
        struct S(T) {}
        alias A(T) = S!T;

        static assert(is(A!int : S!T, T));
        static assert(!is(A!int : A!T, T));
        
---
__stdin.d(5): Error: cannot infer parameters for `is` TypeSpecialization when it is an alias template instance
__stdin.d(5):        while evaluating: `static assert(!is(S!int : A!T, __isexp_id3, T))`
spec/module.dd: 5 examples found
```

dkorpel reviewed (changes requested):
I don't think the issue is valid, see bugzilla

ntrel committed:
Test that an alias parameter does not error

Even when passed an alias template.
This is for metaprogramming as `is` should return false in this case.

ntrel commented:
@dkorpel:
> It's not always false, it's only guaranteed to be false if `A` is an alias template.
is-expressions are used in generic code and supposed to be lenient. Note that this compiles as well:
...
Requiring an extra is-expression to check "am I not passing an alias template to my actual is-expression?" is clumsy.

I have added a test to show that if `A` is itself a template parameter, the error _does not fire_. So are there any other cases where the error fires but could succeed if a different template parameter was passed? I.e. this pull only detects cases that are always a programmer mistake, and should not error on any case where the `is` test could succeed given a different template argument.

dkorpel commented:
There shouldn't be special treatment for parameters, even a global symbol could be behind a `version()` block, behind a `static if`, or the result of refactoring (manual inlining). And if the error doesn't fire for template parameters, it won't catch the common way in which people stumble on the DIP1023 issue, which is from template constraints.

ntrel commented:
@dkorpel 
> even a global symbol could be behind a version() block, behind a static if

Yes, **but the `is` expression is always false regardless**.

> And if the error doesn't fire for template parameters, it won't catch the common way in which people stumble on the DIP1023 issue, which is from template constraints.

This only affects _TypeSpecialization_. The common way is caught:
```d
template t(T)
if (is(T == SomeAliasTemplate!U, U))
```
What I was saying is that SomeAliasTemplate is a global symbol, not a template alias parameter. This usage is always wrong.

dkorpel commented:
> Yes, **but the `is` expression is always false regardless**.

Not when you change the version / static if condition so the symbol isn't an alias template anymore. Also, 'is always false' doesn't imply it should be an error. This is always false as well:
```D
bool f(ubyte x) => x > 300;
```

> The common way is caught:

You're right, my mistake
